### PR TITLE
GA4GH

### PR DESCRIPTION
**Pull request by silverdaz**
This PR implements the GA4GH feature. It is essentially the same architecture as before, but the cryptographic steps are done using the GA4GH standard, and not PGP+custom format for the vault.

Moreover, the vault is backed by an S3 storage.

In more details:
* The inbox is a still a fuse-layer, that stores files to disk. It is _not_ connected at all to the database, nor anything else. Ideally, traffic routes are blocked from the inbox to any other containers/machines.
* The ingestion workers are now lightweight processes that read the first few bytes of the inbox files and stores them in the database. The rest of each file is copied to the vault backend.
* The vault backend is now configurable to either by a file system on disk, or an S3 storage. The default is on disk and the bootstrap is configured to use S3.
* Once the files are split between the database and the vault, a separate process runs a quality control. At the moment, it only checks if it can decrypt the given file and if the checksum is correct. If so, it flags the database entry as completed and a message is communicated to Central EGA. 
Ideally, in the future, many other quality controls can be run as background tasks, without delaying the ingestion procedure. A message can be marked as ingested, but later on, the Quality Control will detect defects in the file, or the file format.
* The keyserver is updated to serve only the PGP keys (since headers are encrypted with OpenPGP). It was not possible to send the keys themselves unlocked over the wire, so the keyserver unlocks them with their appropriate passphrase and relocks them with an "Application Password". The verification step, the (future) quality controllers and the keyserver are the only one to need that password.

The cryptographic part have been moved to the LocalEGA-cryptor repo, as a separate package.

Finally, the bootstrap script does not start multiple instances. It's only one, say lega, along with the cega stubs.

### CODE REVIEW

**Comment by silverdaz** [comment]
@dtitov: Could you please help updating the tests, since there is no "instance" anymore? Containers are simply named `db`, `mq`, `ingest`, `s3`, `verify`...etc...(and `cega-mq`, `cega-users`, `cega-eureka`). So `context.getTargetInstance()` is not necessary.

Unless you want to minimize the changes and just use one instance called `lega`.

**Comment by dtitov** [comment]
@silverdaz, yes, I will.

**Comment by dtitov** [comment]
I've fixed almost all the tests except the ones with successfull encryption scenario, because test-suite at the moment is not able to encrypt files with Crypt4GH. But I'll try to add this functionality tomorrow.

There's one more failing test: `Scenario: A.3 User exists in Central EGA and tries to connect to LocalEGA, but the inbox was not created for him`. Looks like mounting is done somehow differently now, so I have to change simulation of "inbox not created" event, but I'm not sure how...

**Comment by blankdots** [comment]
Unit tests partly re-factored, will continue that

File: lega/utils/db.py [hunk 1]
```diff
@@ -157,49 +153,38 @@ def set_error(file_id, error, from_user=False):
...
-                        {'filepath': filepath, 'file_id': file_id, 'status': Status.Archived.value, 'filesize': filesize})
+                        {'status': Status.Archived.value,
+                         'file_id': file_id,
+                         'vault_path': vault_path,
+                         'vault_filesize': vault_filesize,
+                         'header': header.hex(),
+                        })
 
 ######################################
 ##           Async code             ##
```
Comment (blankdots) [inline lega/utils/db.py:190]: it seems that `async def get_fileinfo(conn, file_id)` is not used anymore or is it just me ?

**Review by blankdots — DISMISSED** [review]

File: lega/utils/db.py (hunk 1 above)
Comment (silverdaz) [reply lega/utils/db.py:190]: We can remove it. Not hard to put it back if necessary.

**Comment by silverdaz** [comment]
@dtitov : Have you seen that: https://github.com/AlexanderSenf/Crypt4GH ?

**Comment by dtitov** [comment]
@silverdaz, yes, it's a draft and Alexander haven't promised that it's production ready. Also, there's no encryption stream there - only decryption stream. And it's not published to Maven as well :(

Anyway, now we have this: https://github.com/uio-bmi/crypt4gh :)

**Comment by blankdots** [comment]
the `lega/conf/loggers` also need to be modified or removed if we are not going to use all of them

**Comment by silverdaz** [comment]
ok, no problem.
But why do the loggers matter?

**Comment by silverdaz** [comment]
I don't mind having another implementation in Java somewhere else.
It's fine.

Just out of curiosity: I think the implementation of Alexander does encryption too.
Am I seeing this wrong?
https://github.com/AlexanderSenf/Crypt4GH/blob/master/src/crypt4gh/Crypt4gh.java#L151-L195

**Comment by dtitov** [comment]
That's correct, it's just not wrapped as an OutputStream.

**Comment by silverdaz** [comment]
and?

**Comment by dtitov** [comment]
And I believe it should be. 

But why are we discussing it here at all? I'm not sure this PR is the right place. I'm fixing the test-suite for this PR, as you asked.

**Comment by silverdaz** [comment]
I'm not discussing.
I'm just pointing out the other implementation to help.

**Comment by dtitov** [comment]
Thank you.

**Comment by silverdaz** [comment]
But for the discussion: I don't think the testsuite should necessarily implement the encryption as users will do. It is not a bad thing to have it, of course, but I think that encrypting a small file and hard-coding it in the testsuite is simple enough, and we can find some other way to test big files.

**Comment by silverdaz** [comment]
crypt4gh is not complicated, so if the testsuite already has encryption, then it's all good!

**Comment by silverdaz** [comment]
Side note: I've got the feeling that we can have very soon another PR with data-out, including a step in the testsuite that decrypts files (as regular https streams or as htsget streams) ;-)

**Comment by dtitov** [comment]
Exactly!

File: deployments/docker/images/base/Dockerfile [hunk 2]
```diff
@@ -3,45 +3,43 @@ LABEL maintainer "NBIS System Developers"
...
+
 # verify the signature
-	export GNUPGHOME="$(mktemp -d)"; \
-	gpg --keyserver ha.pool.sks-keyservers.net --recv-keys B42F6819007F00F88E364FD4036A9C25BF357DD4; \
-	gpg --batch --verify /tmp/gosu.asc /usr/bin/gosu; \
-	rm -r "$GNUPGHOME" /tmp/gosu.asc; \
-	\
-	chmod +x /usr/bin/gosu; \
+RUN export GNUPGHOME="$(mktemp -d)" && \
+    gpg --keyserver ha.pool.sks-keyservers.net --recv-keys B42F6819007F00F88E364FD4036A9C25BF357DD4 && \
```
Comment (blankdots) [inline deployments/docker/images/base/Dockerfile]: @silverdaz can we have this line as:
```
gpg --keyserver ha.pool.sks-keyservers.net --recv-keys "$GPG_KEYS" \
    || gpg --keyserver pgp.mit.edu --recv-keys "$GPG_KEYS" \
    || gpg --keyserver keyserver.pgp.com --recv-keys "$GPG_KEYS" && \
```
 i think it is a firewall issue but `ha.pool.sks-keyservers.net` does not always work
also adding: `ENV GPG_KEYS B42F6819007F00F88E364FD4036A9C25BF357DD4`

Comment (blankdots) [reply deployments/docker/images/base/Dockerfile]: Done :heavy_check_mark:

**Review by blankdots — APPROVED** [review]
Next PR should have docs updated and try to make base image smaller

**Comment by dtitov** [comment]
Tests are green: both locally and on Travis. 

The architecture looks good to me, although I haven't investigated the Python code in details. But since Stefan approved the PR, I'm performing the merge.

**Pull request merged**